File: Java/EjemploSencillo/src/ejemplosencillo/EDA/EDA.java
package ejemplosencillo.EDA;

import ejemplosencillo.EDA.Item;
import ejemplosencillo.base.IFuncionObjetivo;
import java.util.Collections;
import java.util.LinkedList;

public abstract class EDA {
    protected IFuncionObjetivo funcion = null;
    protected Integer Dimension;
    protected Integer Iteraciones;
    protected Integer Poblacion;
    protected Integer Seleccion;
    
    protected LinkedList<Item> poblacion = null;

    public LinkedList<Item> getPoblacion() {
        return poblacion;
    }

    
    public EDA(Integer Dimension, Integer Iteraciones, Integer Poblacion, Integer Seleccion,
            IFuncionObjetivo funcion) {
        this.Dimension = Dimension;
        this.Iteraciones = Iteraciones;
        this.Poblacion = Poblacion;
        this.Seleccion = Seleccion;
        this.funcion = funcion;
    }
    
    public void Ejecutar(){
        try {
            this.poblacion = new LinkedList<>();
            LinkedList<Item> items = Inicializar();
            this.poblacion.add(items.getFirst().clone());
            for (int i=0; i<this.Iteraciones; i++){
                LinkedList<Float> probabilidades = pdf(Seleccion(items, this.Seleccion));
                items = Muestrear(probabilidades, this.Poblacion);
                this.poblacion.add(items.getFirst().clone());
            }
        } catch (Exception ex){
            System.err.println(ex);
        }
    }
    
    private LinkedList<Item> Inicializar(){
        LinkedList<Float> probabilidades = new LinkedList<>();
        for (int i=0; i<this.Dimension; i++){
            probabilidades.add(0.5f);
        }
        
        LinkedList<Item> items = new LinkedList<>();
        for (int i=0; i<this.Poblacion; i++){
            items.add(new Item());
            items.getLast().Generar(probabilidades);
            items.getLast().Evaluar(this.funcion);
        }
        Collections.sort(items);
        return items;
    }
    private LinkedList<Item> Seleccion(LinkedList<Item> items, Integer numer){
        LinkedList<Item> de = new LinkedList<>();
        for (int i=0; i<numer; i++){
            de.add(items.get(i).clone());
        }
        return de;
    }
    public abstract LinkedList<Float> pdf(LinkedList<Item> de);
    public abstract LinkedList<Item> Muestrear(LinkedList<Float> pdf, Integer Num);
    
}

File: Java/MIMIC/src/ejemplosencillo/EDA/EDA.kt
package ejemplosencillo.EDA

import ejemplosencillo.EDA.Item
import ejemplosencillo.base.IFuncionObjetivo
import java.util.Collections
import java.util.LinkedList

abstract class EDA(protected var Dimension: Integer, protected var Iteraciones: Integer, protected var Poblacion: Integer, protected var Seleccion: Integer,
                   funcion: IFuncionObjetivo) {
    protected var funcion: IFuncionObjetivo? = null

    var poblacion: LinkedList<Item>? = null
        protected set


    init {
        this.funcion = funcion
    }

    fun Ejecutar() {
        try {
            this.poblacion = LinkedList()
            var items = Inicializar()
            this.poblacion!!.add(items.getFirst().clone())
            for (i in 0 until this.Iteraciones) {
                val probabilidades = pdf(Seleccion(items, this.Seleccion))
                items = Muestrear(probabilidades, this.Poblacion)
                this.poblacion!!.add(items.getFirst().clone())
            }
        } catch (ex: Exception) {
            System.err.println(ex)
        }

    }

    private fun Inicializar(): LinkedList<Item> {
        val probabilidades = LinkedList()
        for (i in 0 until this.Dimension) {
            probabilidades.add(0.5f)
        }

        val items = LinkedList()
        for (i in 0 until this.Poblacion) {
            items.add(Item())
            items.getLast().Generar(probabilidades)
            items.getLast().Evaluar(this.funcion)
        }
        Collections.sort(items)
        return items
    }

    private fun Seleccion(items: LinkedList<Item>, numer: Integer): LinkedList<Item> {
        val de = LinkedList()
        for (i in 0 until numer) {
            de.add(items.get(i).clone())
        }
        return de
    }

    abstract fun pdf(de: LinkedList<Item>): LinkedList<Float>
    abstract fun Muestrear(pdf: LinkedList<Float>, Num: Integer): LinkedList<Item>

}
